src/AI/WallyAI.java

```diff
@@ -166,29 +225,460 @@ public void endTurn()
+          if( null != readyPlan )
+          {
+            final UnitContext actor = ai.mapPlan[x][y].identity;
+            if( null != actor.unit )
+            {
+              if( actionsBooked.containsKey(actor.unit) )
+                ai.log(String.format("Warning: Unit is double-booked %s:\n\t%s\n\t%s", actor, readyPlan, actionsBooked.get(actor.unit)));
+              else
+                actionsBooked.put(actor.unit, readyPlan);
+            }
+            ai.queuedActions.add(readyPlan);
+            actionAtCoord = movexyc;
+            break;
+          }
+        }
+        if( null != actionAtCoord )
+          revisitTiles.remove(actionAtCoord);
+      }
+
+      // Clear out any stragglers to maybe re-plan
+      for( XYCoord movexyc : revisitTiles )
+      {
+        int x = movexyc.x, y = movexyc.y;
+        final Unit unit = ai.mapPlan[x][y].identity.unit;
+        ai.mapPlan[x][y].identity = null;
+        ai.mapPlan[x][y].toAchieve = null;
+        if( null != unit )
+          ai.plansByUnit.remove(unit);
+      }
+
+      if( ai.queuedActions.isEmpty() )
+        return null;
+
+      ai.log(String.format("  Actions:\n%s", ai.queuedActions));
+      GameAction action = ai.pollAndCleanUpAction(map);
+      return action;
+    }
+
+    private static ActionPlan fetchPlanAndVacateTiles(HashSet<XYCoord> vacatedTiles,
+                                       WallyAI ai, GameMap map,
+                                       int x, int y)
+    {
+      UnitContext actor = ai.mapPlan[x][y].identity;
+      ActionPlan  plan  = ai.mapPlan[x][y].toAchieve;
+      if( null == plan )
+        return null; // Nothing to do here
+
+      final Unit unit = actor.unit;
+      if( null != unit && unit.isTurnOver )
+      {
+        ai.log(String.format("Warning: Action planned for tired unit %s:\n\t%s", unit, plan));
+        return null; // Nothing to do here
+      }
+
+      final XYCoord movexyc = new XYCoord(x, y);
+      if( !vacatedTiles.contains(movexyc)
+          && !map.isLocationEmpty(unit, movexyc)
+          )
+        return null; // Location is full and won't be emptied by our current confirmed plans
+
+      // Assuming our path has been planned well re:terrain, running into enemies is the only obstacle
+      GamePath movePath = plan.path;
+      if( null != movePath )
+      {
+        MoveType fff = actor.calculateMoveType();
+        ArrayList<PathNode> waypoints = movePath.getWaypoints();
+        // We iterate from 1 because the first waypoint is the unit's initial position.
+        for( int i = 1; i < waypoints.size(); i++)
+        {
+          XYCoord from = waypoints.get(i-1).GetCoordinates();
+          XYCoord to   = waypoints.get( i ).GetCoordinates();
+          int cost = fff.getTransitionCost(map, from, to, actor.CO.army, false);
+          if( cost > actor.movePower )
+            if( !vacatedTiles.contains(to) )
+              return null;
+        }
```

Comment:
This iterates over all the waypoints, but line 558 only compares `cost` against `actor.movePower` for the current segment's cost. Should `cost` accumulate across loop iterations? Or does `movePower` not account for distance moved so far? I guess this is just a check to see if the unit can't cross the boundary at all, which should only happen if there is an enemy in the way. If that's the case though, it seems like there should be a more straightforward way to check that.

Reply:
> Should `cost` accumulate across loop iterations?

Arguably, I guess.
> Or does `movePower` not account for distance moved so far? I guess this is just a check to see if the unit can't cross the boundary at all, which should only happen if there is an enemy in the way.

Pretty much.
> If that's the case though, it seems like there should be a more straightforward way to check that.

...Yeah, probably.